design-proposals/shadow-node.md

```diff
@@ -0,0 +1,328 @@
+  | object.(not labels/annotations/resourceVersion/managedFields) == oldObject.(not labels/annotations/resourceVersion/managedFields) | virt-handler user can only change allowed sub-metadata fields.                 |                                                                                                                                                                        |
+  | size(`newNonKubevirtLabels`) == size(`oldNonKubevirtLabels`)                                                                      | virt-handler user cannot add/delete non kubevirt-owned labels                  |                                                                                                                                                                        |
+  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |
+  | !`requestHasNodeName` OR (object.Name == request.nodeName)                                                                        | virt-handler user may only modify nodes that belong to the pod it's running on | needs ServiceAccountTokenNodeBinding feature gate that is [alpha](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L740) on k8s 1.29 |
+
+Pros:
+* Avoid adding CRDs or changing the current virt-handler code to fix the CVE issue.
+* A k8s built-in feature which is straightforward to maintain and configure.
+* validationAdmissionPolicy is [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it could be used for the next Kubevirt minor release.
+
+Cons:
+* The last validation (that prevents virt-handler from writing to other nodes) requires an alpha feature, so it can be excluded until it is GA.
+  * Even without this validation, the CVE risk is greatly reduced.
+
+## Solution option #2: ShadowNode CRD
+The security issue is being addressed by introducing a new light weight CRD: "ShadowNode" for each Kubernetes node.
+Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.
+
+Pros:
+- Since the new controller runs virt-controller context which only runs on the control-plane nodes, the risk of taking over the cluster is reduced.
```

Comment:
Let's rephrase " which only runs on the control-plane nodes,".
The virt-controller is running only on subset of nodes, therefore the exposure is not so wide as the virt-handler. Also for the controller we should highly recommend to run it on control-plane or a nodes that can be more isolated.

Reply:
DONE.